designs/component-upgrades.md

```diff
@@ -0,0 +1,149 @@
+
+Updates to the controller code should also take this into consideration. They always need to consider how to handle existing objects in the cluster. For example, if changing the pattern to name certain CAPI objects in the provider templates, the controller has to be able to handle both the old and the new pattern.
+
+If there is a new eks-a api version, we will need to write conversion webhooks.
+
+### Flux
+
+The process to upgrade the Flux components is very similar to the installation process, it uses the `flux` cli:
+
+1. Write the “kustomization” layer to disk to set our own Flux images
+2. Run the `bootstrap` command. This will generate the new manifest (`flux-system/gotk-components.yaml`) and commit all files to the repository
+3. Run the `reconcile` command for `flux-system` which tells Flux to pull the manifests from git and upgrade itself
+
+This process should work for both new versions and new builds of the same version.
+
+Full documentation: https://fluxcd.io/docs/installation/#bootstrap-upgrade
+
+### Cilium
+
+The process to upgrade cilium is a bit more involved: it uses helm to generate intermediate manifests and kubectl to apply them.
+
+1. Generate pre-flight check with helm and apply it with kubectl. This “pre-pulls" the new images on the nodes to avoid `ErrImagePull` errors later during the upgrade
+2. Once it’s finished, delete the pre-flight components with kubectl
+3. Generate the new Cilium manifest using the right “upgrade compatibility” options with helm (the generated manifest is different depending on the minor version jump)  and apply it with kubectl.
+
+Depending on the implementation, we might be able to avoid creating some/all of these manifests by including them in the bundle manifest.
+
+Full documentation: https://docs.cilium.io/en/v1.10/operations/upgrade/
+
+### CAPI upgrade to `v1alpha4`
+
+As mentioned in the [CAPI](#capi) section, we don’t don’t need any extra process to perform CAPI component upgrades that involve an api version change. However, there will some extra effort needed to support `v1alpha4` in the cli. Most of this will probably apply to future api version changes.
+
+* Add support for `v1alpha4`  in the etcdadm provider. This includes writing conversion webhooks.
+* Port all the patches made to upstream cluster-api to `v1alpha4` CRDs and new controller’s code (build on top of latest patch of `v0.4.x`)
+* Update all provider templates in the cli to generate `v1alpha4` CAPI specs
+* Build compatible versions of dependencies like `clusterctl` and cert-manager
+
+## Customer experience
+
+### Behavior change
+
+There are not backwards breaking changes. However, the expected behavior of the `upgrade cluster` command will be changing, so it should be clearly communicated in the the changelog.
```

Comment:
Let's say we roll out a change and want to revert back to an older release. Will customers be able to run upgrade cluster to get that out or will it always be a roll-forward in our case? For EKS, there are automatic updates, but I believe we also have the case where changes can get "rolled-back" if there is an issue automatically, so curious about the behavior here.

Reply:
Great question
The answer depends on the what's the version change (minor, patch, api version bump, just a new build...).
For example, if we are just reverting back a new build of the components version, yes, we only need a new bundles manifest that points to the previous build. When customers run the upgrade command, the "old" version would be installed.
But, even if this is a downgrade, it's not a rollback per se, it's more a roll-forward like you mentioned. A proper rollback would take the cluster back to the previous installed version, which might not be the version from the previous bundle manifest (image a customer skipping one version upgrade).

For more "drastic" upgrades, like something that involves a CAPI API bump, we would need to test it case by case. Also that would involve a new cli release, since we would need to revert the providers' templates as well.

To be clear, this design assumes that the version changes are always forward. If we want the ability to downgrade versions, we would need to come up with a process for each component (the process for upgrade might work here, but it would need to be tested). If we want the ability to do proper roll-backs, then that also would need a bit more research, since most probably each component would require a slightly different process.